File: 24_10/14/P42584/solution.py
from collections import deque

def solution(prices):
    # 활용할 변수 준비
    N = len(prices)  # 초단위 주식 길이
    stack = deque()  # 연산을 위한 스택
    answer  = [0 for _ in range(N)]  # 정답을 담을 
    
    # 주식가격을 순회하며 넣고 빼기
    idx = 0
    while idx < N:
        if (not stack or stack[-1][0] <= prices[idx]):
            stack.append([prices[idx], idx])
            idx += 1
        else:
            sell = stack.pop()
            answer[sell[1]] = idx - sell[1]
            
    # 남아있는 주식 정리
    idx -= 1
    while stack:
        sell = stack.pop()
        answer[sell[1]] = idx - sell[1]
            
            
    return answer

File: 24_10/14/P42584/solution.js
function solution(prices) {
    // 1. 활용할 변수 정리
    // 1.1 prices의 길이
    const N = prices.length;
    // 1.2 스택 선언([숫자, 인덱스])
    stack = [[prices[0], 0]];
    
    // 2. 스택을 통해서 가격이 떨어지지 않은 기간 구하기
    const answer = new Array(N).fill(20001);
    for (let idx = 1; idx < N; idx++) {
        // 2.1 현재 탐색 중인 주식가격보다 큰 가격이 스택에 존재한다면 모두 꺼내기
        while (stack.length !== 0 && stack[stack.length - 1][0] > prices[idx]) {
            const priceInfo = stack.pop();
            answer[priceInfo[1]] = idx - priceInfo[1];
        }
        // 2.2 스택에 현재 주식 가격 삽입
        stack.push([prices[idx], idx]);
    }
    
    // 3. 스택에 남아있는 값 정리
    while (stack.length !== 0) {
        const priceInfo = stack.pop();
        answer[priceInfo[1]] = N - 1 - priceInfo[1];
    }
    
    return answer;
}

console.log(solution([1, 2, 3, 2, 3]))